pkg/sfu/peer.go

```diff
@@ -74,80 +87,85 @@ func (p *Peer) Join(sid, uid string) error {
 
 	p.session, cfg = p.provider.GetSession(sid)
 
-	p.subscriber, err = NewSubscriber(uid, cfg)
-	if err != nil {
-		return fmt.Errorf("error creating transport: %v", err)
-	}
-	p.publisher, err = NewPublisher(p.session, uid, cfg)
-	if err != nil {
-		return fmt.Errorf("error creating transport: %v", err)
-	}
-
-	for _, dc := range p.session.datachannels {
-		if err := p.subscriber.AddDatachannel(p, dc); err != nil {
-			return fmt.Errorf("error setting subscriber default dc datachannel")
-		}
-	}
-
-	p.subscriber.OnNegotiationNeeded(func() {
-		p.Lock()
-		defer p.Unlock()
-
-		if p.remoteAnswerPending {
-			p.negotiationPending = true
-			return
-		}
-
-		Logger.V(1).Info("Negotiation needed", "peer_id", p.id)
-		offer, err := p.subscriber.CreateOffer()
+	if !conf.NoSubscribe {
+		p.subscriber, err = NewSubscriber(uid, cfg)
 		if err != nil {
-			Logger.Error(err, "CreateOffer error")
-			return
-		}
-
-		p.remoteAnswerPending = true
-		if p.OnOffer != nil && !p.closed.get() {
-			Logger.V(0).Info("Send offer", "peer_id", p.id)
-			p.OnOffer(&offer)
+			return fmt.Errorf("error creating transport: %v", err)
 		}
-	})
 
-	p.subscriber.OnICECandidate(func(c *webrtc.ICECandidate) {
-		Logger.V(1).Info("On subscriber ice candidate called for peer", "peer_id", p.id)
-		if c == nil {
-			return
-		}
-
-		if p.OnIceCandidate != nil && !p.closed.get() {
-			json := c.ToJSON()
-			p.OnIceCandidate(&json, subscriber)
-		}
-	})
+		p.subscriber.OnNegotiationNeeded(func() {
+			p.Lock()
+			defer p.Unlock()
+
+			if p.remoteAnswerPending {
+				p.negotiationPending = true
+				return
+			}
+
+			Logger.V(1).Info("Negotiation needed", "peer_id", p.id)
+			offer, err := p.subscriber.CreateOffer()
+			if err != nil {
+				Logger.Error(err, "CreateOffer error")
+				return
+			}
+
+			p.remoteAnswerPending = true
+			if p.OnOffer != nil && !p.closed.get() {
+				Logger.V(0).Info("Send offer", "peer_id", p.id)
+				p.OnOffer(&offer)
+			}
+		})
+
+		p.subscriber.OnICECandidate(func(c *webrtc.ICECandidate) {
+			Logger.V(1).Info("On subscriber ice candidate called for peer", "peer_id", p.id)
+			if c == nil {
+				return
+			}
+
+			if p.OnIceCandidate != nil && !p.closed.get() {
+				json := c.ToJSON()
+				p.OnIceCandidate(&json, subscriber)
+			}
+		})
+	}
 
-	p.publisher.OnICECandidate(func(c *webrtc.ICECandidate) {
-		Logger.V(1).Info("on publisher ice candidate called for peer", "peer_id", p.id)
-		if c == nil {
-			return
+	if !conf.NoPublish {
+		p.publisher, err = NewPublisher(p.session, uid, cfg)
+		if err != nil {
+			return fmt.Errorf("error creating transport: %v", err)
 		}
-
-		if p.OnIceCandidate != nil && !p.closed.get() {
-			json := c.ToJSON()
-			p.OnIceCandidate(&json, publisher)
+		for _, dc := range p.session.datachannels {
+			if err := p.subscriber.AddDatachannel(p, dc); err != nil {
```

Comment:
As far as I understand, there will always be an error in the case of `conf.NoSubscribe == true`

Reply:
You are right!

Reply:
In general, it turned out that with this new conf, the "two pc system" (https://github.com/pion/ion-sdk-js/commit/6dff082093893ae7befc1317e78cf05867faa464) in the ion-sdk-js changed a lot.